src/components/Notifications.js

```diff
@@ -16,14 +26,159 @@ const defaultNotificationProps = {
+        cursor: 'auto',
+        display: 'flex',
+        flexDirection: 'column',
+        fontSize: 12
+      }
+    }, [
+      // content and close button
+      div({ style: { display: 'flex', padding: '0.75rem 1rem' } }, [
+        // content
+        div({ style: { display: 'flex', flex: 1, flexDirection: 'column' } }, [
+          // icon and title
+          div({ style: { alignItems: 'center', display: 'flex' } }, [
+            !!iconType && icon(iconType, { className: 'is-solid', size: 26, style: { marginRight: '0.5rem' } }),
+            div({ style: { fontWeight: 600 } }, [title])
+          ]),
+          !!message && div({ style: { marginTop: '0.5rem' } }, [message]),
+          !!detail && h(Clickable, {
+            style: { marginTop: '0.25rem', textDecoration: 'underline' },
+            onClick: () => this.setState({ modal: true })
+          }, ['Details'])
+        ]),
+        h(Clickable, {
+          title: 'Dismiss notification',
+          onClick: () => notificationsRef.current.removeNotification(id)
+        }, [icon('times', { size: 20 })])
+      ]),
+      notifications.length > 1 && div({ style: { alignItems: 'center', borderTop: `1px solid ${color[1]}`, display: 'flex', fontSize: 10, padding: '0.75rem 1rem' } }, [
+        h(Clickable, {
+          disabled: onFirst,
+          style: { color: onFirst ? color[1] : null },
+          onClick: () => this.setState({ notificationNumber: notificationNumber - 1 })
+        }, [icon('angle left', { size: 12 })]),
+        div({
+          style: {
+            backgroundColor: color[1],
+            borderRadius: 10,
+            padding: '0.2rem 0.5rem'
+          }
+        }, [
+          notificationNumber + 1, '/', notifications.length
+        ]),
+        h(Clickable, {
+          disabled: onLast,
+          style: { color: onLast ? color[1] : null },
+          onClick: () => this.setState({ notificationNumber: notificationNumber + 1 })
+        }, [icon('angle right', { size: 12 })])
+      ]),
+      modal && h(Modal, {
+        width: 800,
+        title,
+        showCancel: false,
+        showX: true,
+        onDismiss: () => this.setState({ modal: false }),
+        okButton: buttonPrimary({ onClick: () => refreshPage() }, 'Refresh Page')
+      }, [
+        h(ErrorView, { error: detail, collapses: false })
+      ])
+    ])
+  }
+})
+
+const refreshPage = () => {
+  notificationStore.set([])
+  StateHistory.clearCurrent()
+  // document.location.reload()
+}
+
+const showNotification = ({ id, timeout }) => {
+  notificationsRef.current.addNotification(_.merge(defaultNotificationProps, {
+    id,
+    content: div({ style: { width: '100%' } }, [
+      h(NotificationDisplay, { id })
+    ]),
+    dismiss: { duration: !!timeout ? timeout : 0 },
+    dismissable: { click: false, touch: false },
+    width: 350
+  }))
+}
+
+class Notifications extends Component {
+  static getDerivedStateFromProps({ notificationState }, { visibleNotifications = [] }) {
```

Comment:
Since this is performing side effects, it would be preferable to do this in a regular lifecycle method, e.g. `componentDidUpdate`. I think we'd be able to compute the difference between the previous and current states.

Reply:
I'd have to set state inside `componentDidUpdate`. Wouldn't that risk causing an infinite loop of updates? I'm currently running into this same problem while putting the finishing touches on the Contact Us form changes (making sure the contact email is populated now that the component's lifecycle is different... but now that I think about it, I may be able to solve that another way).

Reply:
You mean `visibleNotifications`? I was thinking that we don't actually need to store that, if we consistently launch any new ids on each update.

Reply:
Update: I did get it to work on Friday.